File: z/merge_sort.cc
// https://app.laicode.io/app/problem/9

class Solution {
 public:
  vector<int> mergeSort(vector<int> array) {
    // write your solution here
    if (array.empty()) {
      return array;
    }
    vector<int> helper(array.size());
    sort(array, 0, array.size() - 1, helper);
    return array;
  }

  void sort(vector<int>& array, int start, int end, vector<int>& helper) {
    if (start >= end) {
      return;
    }
    int mid = start + (end - start) / 2;
    sort(array, start, mid, helper);
    sort(array, mid + 1, end, helper);
    merge(array, start, end, helper);
  }

  void merge(vector<int>& array, int start, int end, vector<int>& helper) {
    int mid = start + (end - start) / 2;
    int left = start;
    int right = mid + 1;
    int index = left;
    while (left <= mid && right <= end) {
      if (array[left] <= array[right]) {
        helper[index++] = array[left++];
      } else {
        helper[index++] = array[right++];
      }
    }
    while (left <= mid) {
      helper[index++] = array[left++];
    }
    while (right <= end) {
      helper[index++] = array[right++];
    }
    for (int i = start; i <= end; ++i) {
      array[i] = helper[i];
    }
  }
};

File: h/mergeSort.java
public class Solution {
    public int[] mergeSort(int[] array) {
      if (array.length <= 1) {
        return array;
      }
  
      int [] helper = new int[array.length];
      sort(array, helper, 0, array.length - 1);
      return array;
    }
  
    private void sort(int [] array, int [] helper, int left, int right) {
      if (left >= right) {
        return;
      }
  
      int mid = left + (right - left) / 2;
      sort(array, helper, left, mid);
      sort(array, helper, mid + 1, right);
      merge(array, helper, left, mid, right);
    }
  
    private void merge(int [] array, int [] helper, int left, int mid, int right) {
      for (int i = left; i <= right; i++) {
              helper[i] = array[i];
          }
  
          int i = left;
          int j = mid + 1;
          int idx = left;
          while (i <= mid && j <= right) {
              if (helper[i] < helper[j]) {
                  array[idx] = helper[i];
                  idx++;
                  i++;
              } else {
                  array[idx] = helper[j];
                  idx++;
                  j++;
              }
          }
  
          while (i <= mid) {
              array[idx++] = helper[i++];
          } 
    }
  }
